File: app/src/main/java/com/czbix/v2ex/ui/loader/TopicListLoader.java
package com.czbix.v2ex.ui.loader;

import android.content.Context;

import com.czbix.v2ex.dao.TopicDao;
import com.czbix.v2ex.model.Page;
import com.czbix.v2ex.model.Topic;
import com.czbix.v2ex.network.RequestHelper;

import java.util.List;

public class TopicListLoader extends AsyncTaskLoader<List<Topic>> {
    private final Page mPage;

    public TopicListLoader(Context context, Page page) {
        super(context);

        mPage = page;
    }

    @Override
    public List<Topic> loadInBackgroundWithException() throws Exception {
        final List<Topic> topics = RequestHelper.getTopics(mPage);
        for (Topic topic : topics) {
            final int lastRead = TopicDao.getLastRead(topic.getId());
            if (lastRead >= topic.getReplyCount()) {
                topic.setHasRead();
            }
        }
        return topics;
    }
}

File: app/src/main/kotlin/com/czbix/v2ex/ui/loader/TopicListLoader.kt
package com.czbix.v2ex.ui.loader

import android.content.Context
import androidx.collection.ArraySet
import com.czbix.v2ex.db.TopicRecordDao
import com.czbix.v2ex.model.Page
import com.czbix.v2ex.model.Topic
import com.czbix.v2ex.network.RequestHelper
import kotlinx.coroutines.runBlocking

class TopicListLoader(
        context: Context,
        private val mPage: Page,
        private val dao: TopicRecordDao
) : AsyncTaskLoader<TopicListLoader.TopicList>(context) {
    @Throws(Exception::class)
    override fun loadInBackgroundWithException(): TopicList {
        val topics = runBlocking {
            RequestHelper.getTopics(mPage)
        }

        val readed = topics.filter { topic ->
            val lastRead = dao.getLastReadComment(topic.id) ?: 0
            lastRead >= topic.replyCount
        }.map {
            it.id
        }

        topics.readed = ArraySet(readed)

        return topics
    }

    class TopicList(
            list: List<Topic>,
            val isFavorited: Boolean,
            val onceToken: String? = null
    ): List<Topic> by list {
        lateinit var readed: Set<Int>
    }
}
